src/Analysis/Ast/Impl/Analyzer/PythonAnalyzer.cs

```diff
@@ -14,145 +14,366 @@
+            try {
+                var startTime = stopWatch.ElapsedMilliseconds;
+                var module = node.Value.Module;
+                var ast = node.Value.Ast;
+
+                // Now run the analysis.
+                var walker = new ModuleWalker(_services, module, ast);
+
+                await ast.WalkAsync(walker, cancellationToken);
+                cancellationToken.ThrowIfCancellationRequested();
 
-            using (var cts = CancellationTokenSource.CreateLinkedTokenSource(_globalCts.Token, cancellationToken)) {
-                var dependencyRoot = await _dependencyResolver.GetDependencyChainAsync(document, cts.Token);
-                // Notify each dependency that the analysis is now pending
-                NotifyAnalysisPending(document, dependencyRoot);
+                // Note that we do not set the new analysis here and rather let
+                // Python analyzer to call NotifyAnalysisComplete.
+                await walker.CompleteAsync(cancellationToken);
+                cancellationToken.ThrowIfCancellationRequested();
+                var analysis = new DocumentAnalysis((IDocument)module, version, walker.GlobalScope, walker.Eval);
+
+                (module as IAnalyzable)?.NotifyAnalysisComplete(analysis);
+                lock (_syncObj) {
+                    node.Value.TrySetAnalysis(analysis, version);
+                }
 
-                cts.Token.ThrowIfCancellationRequested();
-                await AnalyzeChainAsync(dependencyRoot, cts.Token);
+                node.MarkCompleted();
+                _log?.Log(TraceEventType.Verbose, $"Analysis of {module.Name}({module.ModuleType}) complete in {stopWatch.ElapsedMilliseconds - startTime} ms.");
+            } catch (OperationCanceledException oce) {
+                lock (_syncObj) {
+                    node.Value.TryCancel(oce, version);
+                }
+            } catch (Exception exception) {
+                lock (_syncObj) {
+                    node.Value.TrySetException(exception, version);
+                }
+            } finally {
+                Interlocked.Decrement(ref _runningTasks);
             }
         }
 
-        private void NotifyAnalysisPending(IDocument document, IDependencyChainNode node) {
-            // Notify each dependency that the analysis is now pending except the source
-            // since if document has changed, it already incremented its expected analysis.
-            if (node.Analyzable != document) {
-                node.Analyzable.NotifyAnalysisPending();
+        [DebuggerDisplay("{Name} : {FilePath}")]
+        private struct ModuleKey : IEquatable<ModuleKey> {
+            public string Name { get; }
+            public string FilePath { get; }
+            public bool IsTypeshed { get; }
+
+            public ModuleKey(IPythonModule module) {
+                Name = module.Name;
+                FilePath = module.ModuleType == ModuleType.CompiledBuiltin ? null : module.FilePath;
```

Comment:
I think just `compiled` won't have file path either. I.e. compiled libraries that are not part of the language library set.

Reply:
AFAIK, we have files like `.pyd` or `.dylib` for `Compiled`

Reply:
OK